File: www/src/models/node.ts
import { Data } from ".";
import { Topic } from "./topic";
import { Service } from "./service";
import { Action } from "./action";

export type Node = {
  name: string;
  namespace: string;
  fullName: string;
  interface?: NodeInterface;
} & Data;

export const nodeKeys = ["name", "namespace", "fullName"] as const;
export type NodeKeyType = typeof nodeKeys[number];
export function isNodeKey(key: string): key is NodeKeyType {
  return nodeKeys.includes(key as NodeKeyType);
}

export type NodeInterface = {
  nodeName: string;
  subscribers?: Topic[];
  publishers?: Topic[];
  serviceServers?: Service[];
  serviceClients?: Service[];
  actionServers?: Action[];
  actionClients?: Action[];
};

File: ros2web/api/models/node.py
from dataclasses import dataclass, field
from typing import Optional
from typing import List, Any

from .topic import Topic
from .service import Service
from .action import Action
from .data import Data

@dataclass
class NodeInterface:
    node_name: str
    subscribers: Optional[List[Topic]] = field(default=None)
    publishers: Optional[List[Topic]] = field(default=None)
    service_servers: Optional[List[Service]] = field(default=None)
    service_clients: Optional[List[Service]] = field(default=None)
    action_servers: Optional[List[Action]] = field(default=None)
    action_clients: Optional[List[Action]] = field(default=None)
    
@dataclass
class Node(Data):
    name: str
    namespace: str
    full_name: str
    interface: Optional[NodeInterface] = field(default=None)
    